src/Workspaces/Remote/Core/SolutionAssetProvider.cs

```diff
@@ -52,58 +72,144 @@ public async ValueTask GetAssetsAsync(PipeWriter pipeWriter, Checksum solutionCh
+                => await _writer.FlushAsync(cancellationToken).ConfigureAwait(false);
+
+            public override void Write(byte[] buffer, int offset, int count)
+            {
+                Requires.NotNull(buffer, nameof(buffer));
+                Verify.NotDisposed(this);
+
+                var span = _writer.GetSpan(count);
+                buffer.AsSpan(offset, count).CopyTo(span);
+                _writer.Advance(count);
+            }
+
+            public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                this.Write(buffer, offset, count);
+                return Task.CompletedTask;
+            }
+
+            public override void WriteByte(byte value)
+            {
+                Verify.NotDisposed(this);
+                var span = _writer.GetSpan(1);
+                span[0] = value;
+                _writer.Advance(1);
+            }
+
+#if !NETSTANDARD
+
+            public override void Write(ReadOnlySpan<byte> buffer)
             {
-                try
-                {
-                    var stream = localPipe.Writer.AsStream(leaveOpen: false);
-                    using var writer = new ObjectWriter(stream, leaveOpen: false, cancellationToken);
-                    RemoteHostAssetSerialization.WriteData(writer, singleAsset, assetMap, serializer, scope.ReplicationContext, solutionChecksum, checksums, cancellationToken);
-                }
-                catch (Exception e) when (FatalError.ReportAndCatchUnlessCanceled(e, cancellationToken))
-                {
-                    // no-op
-                }
-            }, mustNotCancelToken);
-
-            // Complete RPC once we send the initial piece of data and start waiting for the writer to send more,
-            // so the client can start reading from the stream. Once CopyPipeDataAsync completes the pipeWriter
-            // the corresponding client-side pipeReader will complete and the data transfer will be finished.
-            var task2 = CopyPipeDataAsync();
-
-            await Task.WhenAll(task1, task2).ConfigureAwait(false);
-
-            async Task CopyPipeDataAsync()
+                Verify.NotDisposed(this);
+                var span = _writer.GetSpan(buffer.Length);
+                buffer.CopyTo(span);
+                _writer.Advance(buffer.Length);
+            }
+
+            public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                this.Write(buffer.Span);
+                return default;
+            }
+
+#endif
+
+            protected override void Dispose(bool disposing)
+            {
+                this.IsDisposed = true;
+                base.Dispose(disposing);
+            }
+
```

Comment:
Nit: move below constructor, so that construction and disposal are closer together.

Reply:
done.